### ISSUE TITLE
Cargo-powered diagnostics fail silently when a dependency fails to build

### ISSUE THREAD

**Issue by teozkr** [rust-analyzer/rust-analyzer#7517]
# Repro

1. Set up a new project with broken dependencies (assuming cargo-edit and cargo-feature, otherwise do the Cargo.toml edits manually :P):

```bash
cargo new my-repro --bin
cd my-repro
cargo add k8s-openapi@0.11.0 --no-default-features
# Only one version feature may be enabled at a time
cargo feature k8s-openapi +v1_16 +v1_17
# Confirm that it fails to build...
cargo build
``` 

2. Open in whatever Rust-Analyzer-powered editor you prefer (so far I've tested and confirmed emacs+lsp-mode and vscode+ra plugin)
3. Write some invalid code (such as adding a call to a missing `asdf();` function to `main`)

# Expected outcome

Rust-Analyzer reports an error building dependencies.

# Actual outcome

Rust-Analyzer reports no errors, and displays the `asdf();` call as a valid call of type `{unknown}`.

# Well, duh, of course this doesn't work

This example is pretty contrived, but it's easy to get bit by in multi-crate workspaces. If you have multiple `bin` crates then they can end up with conflicting dependencies, which work fine when building any crate in isolation (such as running `cargo run`, or `cargo build` inside of the crate directory), but break when building the whole workspace (such as running `cargo build` in the workspace dir, or running Rust-Analyzer).

**Comment by theduke** [comment]
I have also encountered this regularly.

The issue started happening on master 1-2 weeks ago. 

Note: this is not exclusive to dependencies!
I also see this happening inside a single crate between modules.

It's extremely cumbersome during refactoring, since the issue seems to pop up especially when changing a lot (type signatures, moving modules around, ...).

**Comment by teozkr** [comment]
@theduke 

Just to confirm that we're talking about the same issue... Are you talking about `{unknown}` showing up, or errors from `rustc`/`cargo build` missing in Rust-Analyzer? This issue also seems to be completely deterministic: it's unaffected by restarting Rust-Analyzer, the buffer, or the editor.

**Comment by theduke** [comment]
> Are you talking about {unknown} showing up, or errors from rustc/cargo build missing in Rust-Analyzer?

The latter. 

> This issue also seems to be completely deterministic

This is the behaviour I'm seeing as well.

Things start to work again after fixing enough of the errors, and stop working when undoing those changes.

If it's helpful I can also try to find a minimal single-crate reproduction.

**Comment by lnicola** [comment]
This is more or less working after #8355:

![image](https://user-images.githubusercontent.com/308347/113886818-a3f69900-97c9-11eb-8faa-ed002ce77c7c.png)

I'll file a new issue for the ungreat error messages.
